### PR TITLE
step-83: checkpoint/restart and serialization

### PR DESCRIPTION
**Pull request by bangerth**
Nothing here so far other than a mostly complete introduction. This is going to take a while to be ready.

c.f. #11998

### CODE REVIEW

File: examples/step-83/doc/intro.dox
Comment (anne-glerum) [inline examples/step-83/doc/intro.dox:291]:
```suggestion
should also depend on where in the program's algorithm one
```

Comment (anne-glerum) [reply examples/step-83/doc/intro.dox:291]: Just reading through this step out of interest, hope you don't mind the suggestion :)

Comment (bangerth) [reply examples/step-83/doc/intro.dox:291]: Quite the contrary! Thank you!

**Comment by peterrum** [comment]
@bangerth Will you be able to make it until the release?

**Comment by bangerth** [comment]
@pcafrica This is the program I mentioned. The code itself is currently only a copy of step-19, but good parts of the introduction and results section are already there.

**Comment by drwells** [comment]
@bangerth My students have been asking me a bunch of questions about serialization recently - I forgot that we have this tutorial in a half-finished state. Do you have time to work on this in the next week or two? Would you like to collaborate?

**Comment by pcafrica** [comment]
Hi @drwells!

During the last developer workshop @blaisb and I have been working on this tutorial.

The tutorial is almost complete, but it still depends on PR #15706, which has been open for quite a long time now - after all, it is a huge refractoring.

I would be happy to work on the review - and definitely any help is welcome! 😊

**Comment by blaisb** [comment]
Once the merge of the solution transfer is finished, I can finish the example in a day or less. I'm just waiting for this PR.

**Comment by bangerth** [comment]
@pcafrica and @blaisb Do you have commits that are not in this PR? All I can see here is what I contributed originally, but nothing from you two. If you can give me what you have, I can start reviewing and adding what is still missing.

**Comment by pcafrica** [comment]
@bangerth the most up-to-date version @blaisb and I have worked on is in my forked branch: https://github.com/pcafrica/dealii/tree/step-83

I don't think I have the permission to push here, but maybe @blaisb can or you can cherry-pick?

**Comment by bangerth** [comment]
@pcafrica  Thanks for the link. I'll work on merging the two branches. This is going to take a while, for some reason your branch also has some stuff in the step-86 directory. But I'll figure it out!

**Comment by bangerth** [comment]
@pcafrica and @blaisb I've looked through the current state, and realize that everything is actually in place already! I like this a lot, and once we have #17250 fixed, I think that this can be made ready within a day or two. I worked myself through what I had, what you added, and got it pretty close I think.

It's only sequential, and does not require any of the `SolutionTransfer` stuff at all, but that is ok -- it's a nice first step to learn about serialization, and we can leave the parallel case for a separate program at a later time. So my vote is to just complete what we have here, get it merged, and deal with the parallel case after the release. Does that sound reasonable to you?

**Comment by pcafrica** [comment]
Hi @bangerth!

I really like how the tutorial looks! I think this will be very useful and also very easy to follow.

I do agree with you: since this is basically done, let's merge this PR for the _serial serialization_ (:smile:), and then after the release (hopefully during the workshop) we can keep working on the parallel case.

Thanks a lot.

**Comment by bangerth** [comment]
I will update once #17252 is merged. 

There is one more thing I want to change: Right now, we have `checkpoint()` and `restart()` functions that have both the logic for (de)serialization and for writing/reading checkpoint files. I would like to separate these into a `serialize()` function, and then just `checkpoint()`/`restart()`. I think this is good practice so that if there is a class at a higher level (say, if one wanted to couple this program with another), it can take over the checkpointing/restarting functionality itself and just serialize the object like one would with any other object. This should be relatively straightforward once #17252 is merged.

**Comment by blaisb** [comment]
> I will update once #17252 is merged.
> 
> There is one more thing I want to change: Right now, we have `checkpoint()` and `restart()` functions that have both the logic for (de)serialization and for writing/reading checkpoint files. I would like to separate these into a `serialize()` function, and then just `checkpoint()`/`restart()`. I think this is good practice so that if there is a class at a higher level (say, if one wanted to couple this program with another), it can take over the checkpointing/restarting functionality itself and just serialize the object like one would with any other object. This should be relatively straightforward once #17252 is merged.

Sure, that's a great idea.
@bangerth What's your roadmap for this and where do you need help? It seems like in this current version the tutorial is very close to being finished. Do you need reviewers or you want me or @pcafrica to do things in particular? I'd be glad to help.

**Comment by bangerth** [comment]
I will be on vacation from tomorrow to next Tuesday. I hope to get things wrapped up from my side today -- let me work on it, and then I can put forward a complete version. Then you and everyone have a couple of days to look that through, comment on, create patches again (you can create pull requests on pull requests), and/or approve/merge. Sound good?

**Comment by blaisb** [comment]
> I will be on vacation from tomorrow to next Tuesday. I hope to get things wrapped up from my side today -- let me work on it, and then I can put forward a complete version. Then you and everyone have a couple of days to look that through, comment on, create patches again (you can create pull requests on pull requests), and/or approve/merge. Sound good?

Perfect
I'll review it while you are on vacation and instead of commenting everywhere, if I have larger things I'll make a PR on your PR. Overall I think this is looking good though. I remembered the tutorial being very very closed to being finished.

**Comment by bangerth** [comment]
@blaisb @pcafrica I've got it as far as I can. I would appreciate if you could take it from here.

I will note that there is one thing that does not currently work, but I have not tried to debug this. We will have to fix that before the program is ready. Specifically, if you interrupt at, say, time step 11, you get
```
Timestep 10
  Field degrees of freedom:                 4989
  Total number of particles in simulation:  186
  Number of particles lost this time step:  0

  Now at t=1.03349e-06, dt=6.56398e-08.

--- Writing checkpoint... ---

***************  Total number of particles in simulation:  186
Timestep 11
  Field degrees of freedom:                 4989
  Total number of particles in simulation:  198
  Number of particles lost this time step:  0

  Now at t=1.11482e-06, dt=8.13268e-08.

^C
```
Note the debug output starting with `*****` (which we should remove before merging) that says we have 186 particles. When you then do `./step-83 restart`, you end up with this:
```
--- Restarting at t=1.03349e-06, dt=6.56398e-08

***************  Total number of particles in simulation:  186
Timestep 11
  Field degrees of freedom:                 4989
  Total number of particles in simulation:  20
  Number of particles lost this time step:  20
  Fraction of particles lost through anode: 0

  Now at t=0.0001, dt=9.89665e-05.
```
So it immediately seems to forget all but 20 particles, seems to forget these too, and then takes an infinite time step to the end time. I bet it's something silly we're doing, but this is as far as I can take it before I have to leave it. Any help appreciated! 

(I think that other than this one issue, and then having to add the missing piece in `results.dox` that shows that we get the same output as in a from-scratch run, everything else is in essence ready.)

**Comment by pcafrica** [comment]
@bangerth @blaisb I think this has to do with how `ParticleHandler::serialize` is implemented: https://www.dealii.org/current/doxygen/deal.II/particle__handler_8h_source.html#l01319

**Comment by bangerth** [comment]
Ah yes, that is awkward. How do we address this? What's the right way to serialize particles then? Or should we just fix the `ParticleHandler` function and also serialize particles at least if the underlying triangulation is sequential?

**Comment by peterrum** [comment]
In https://github.com/peterrum/dealii/commits/step-83/, I rebased this PR on https://github.com/dealii/dealii/pull/15706 and made some small changes in the turial as @blaisb 
 suggested and removed some old asserts (https://github.com/peterrum/dealii/commit/35e9dbbfade8e242c04b1b1d485843cda4d23760)
```diff
diff --git a/examples/step-83/step-83.cc b/examples/step-83/step-83.cc
index 76cd0d2a19..f7f7eea504 100644
--- a/examples/step-83/step-83.cc
+++ b/examples/step-83/step-83.cc
@@ -714,8 +714,8 @@ namespace Step83
   void CathodeRaySimulator<dim>::serialize(Archive &ar,
                                            const unsigned int /* version */)
   {
-    ar &triangulation;
-    ar &solution;
+    // ar &triangulation;
+    // ar &solution;
     ar &particle_handler;
     ar &next_unused_particle_id;
     ar &n_recently_lost_particles;
@@ -748,6 +748,9 @@ namespace Step83
   {
     std::cout << "--- Writing checkpoint... ---" << std::endl << std::endl;
 
+    particle_handler.prepare_for_serialization();
+    triangulation.save("abba");
+
     {
       std::ofstream checkpoint_file("checkpoint.tmp");
       AssertThrow(checkpoint_file,
@@ -794,6 +797,8 @@ namespace Step83
   template <int dim>
   void CathodeRaySimulator<dim>::restart()
   {
+    triangulation.load("abba");
+
     std::ifstream checkpoint_file("checkpoint");
     AssertThrow(checkpoint_file,
                 ExcMessage("Could not read from the <checkpoint> file."));
@@ -801,6 +806,8 @@ namespace Step83
     boost::archive::text_iarchive archive(checkpoint_file);
     archive >> *this;
 
+    particle_handler.deserialize();
+
     std::cout << "--- Restarting at t=" << time.get_current_time()
               << ", dt=" << time.get_next_step_size() << std::endl
               << std::endl;
diff --git a/source/particles/particle_handler.cc b/source/particles/particle_handler.cc
index 1e9d71ef67..82e780ee90 100644
--- a/source/particles/particle_handler.cc
+++ b/source/particles/particle_handler.cc
@@ -2194,20 +2194,21 @@ namespace Particles
   void
   ParticleHandler<dim, spacedim>::register_data_attach()
   {
-    parallel::DistributedTriangulationBase<dim, spacedim>
-      *distributed_triangulation =
-        const_cast<parallel::DistributedTriangulationBase<dim, spacedim> *>(
-          dynamic_cast<
-            const parallel::DistributedTriangulationBase<dim, spacedim> *>(
-            &(*triangulation)));
-    (void)distributed_triangulation;
-
-    Assert(
-      distributed_triangulation != nullptr,
-      ExcMessage(
-        "Mesh refinement in a non-distributed triangulation is not supported "
-        "by the ParticleHandler class. Either insert particles after mesh "
-        "creation and do not refine afterwards, or use a distributed triangulation."));
+    // parallel::DistributedTriangulationBase<dim, spacedim>
+    //   *distributed_triangulation =
+    //     const_cast<parallel::DistributedTriangulationBase<dim, spacedim> *>(
+    //       dynamic_cast<
+    //         const parallel::DistributedTriangulationBase<dim, spacedim> *>(
+    //         &(*triangulation)));
+    //(void)distributed_triangulation;
+    //
+    // Assert(
+    //   distributed_triangulation != nullptr,
+    //   ExcMessage(
+    //     "Mesh refinement in a non-distributed triangulation is not supported
+    //     " "by the ParticleHandler class. Either insert particles after mesh "
+    //     "creation and do not refine afterwards, or use a distributed
+    //     triangulation."));
 
     const auto callback_function =
       [this](const typename Triangulation<dim, spacedim>::cell_iterator
@@ -2216,8 +2217,9 @@ namespace Particles
         return this->pack_callback(cell_iterator, cell_status);
       };
 
-    handle = distributed_triangulation->register_data_attach(
-      callback_function, /*returns_variable_size_data=*/true);
+    handle = const_cast<Triangulation<dim, spacedim> *>(&*triangulation)
+               ->register_data_attach(callback_function,
+                                      /*returns_variable_size_data=*/true);
   }
 
 
@@ -2256,20 +2258,21 @@ namespace Particles
   ParticleHandler<dim, spacedim>::notify_ready_to_unpack(
     const bool serialization)
   {
-    parallel::DistributedTriangulationBase<dim, spacedim>
-      *distributed_triangulation =
-        const_cast<parallel::DistributedTriangulationBase<dim, spacedim> *>(
-          dynamic_cast<
-            const parallel::DistributedTriangulationBase<dim, spacedim> *>(
-            &(*triangulation)));
-    (void)distributed_triangulation;
-
-    Assert(
-      distributed_triangulation != nullptr,
-      ExcMessage(
-        "Mesh refinement in a non-distributed triangulation is not supported "
-        "by the ParticleHandler class. Either insert particles after mesh "
-        "creation and do not refine afterwards, or use a distributed triangulation."));
+    // parallel::DistributedTriangulationBase<dim, spacedim>
+    //   *distributed_triangulation =
+    //     const_cast<parallel::DistributedTriangulationBase<dim, spacedim> *>(
+    //       dynamic_cast<
+    //         const parallel::DistributedTriangulationBase<dim, spacedim> *>(
+    //         &(*triangulation)));
+    //(void)distributed_triangulation;
+    //
+    // Assert(
+    //   distributed_triangulation != nullptr,
+    //   ExcMessage(
+    //     "Mesh refinement in a non-distributed triangulation is not supported
+    //     " "by the ParticleHandler class. Either insert particles after mesh "
+    //     "creation and do not refine afterwards, or use a distributed
+    //     triangulation."));
 
     // First prepare container for insertion
     clear();
@@ -2293,8 +2296,8 @@ namespace Particles
             this->unpack_callback(cell_iterator, cell_status, range_iterator);
           };
 
-        distributed_triangulation->notify_ready_to_unpack(handle,
-                                                          callback_function);
+        const_cast<Triangulation<dim, spacedim> *>(&*triangulation)
+          ->notify_ready_to_unpack(handle, callback_function);
 
         // Reset handle and update global numbers.
         handle = numbers::invalid_unsigned_int;
```
 and it works. Just the last patch is not enough, since data has to be copied around within the `Triangulation` class.

**Comment by peterrum** [comment]
What I also noticed is that the serialization of solution vector, at the moment, is not needed (and I have removed it).

I would suggest to change the tutorial in such a way that the loaded vector is actually needed. Furthermore, loading the vector before we setup the system is not how we do it in practice and should not shown in this way. Serialization of vectors in a too important topic (which has costed us all too much time to get it right the first time, since we needed to restructure our code)!

**Comment by bangerth** [comment]
@peterrum I'm on vacation and haven't looked at your patches yet, but as an alternative, have thought of the following:
* We add the serialization of the missing variables in `ParticleHandler` *if the underlying triangulation is sequential*.
* We call it a day on the tutorial program, and make clear that it is only intended to be used for sequential situations -- in that case, serializing the vector as is currently done is appropriate.
* I add a paragraph in the "possibilities for extensions" part of the results section that talks about why checkpointing parallel applications is more complicated.
* We put your approach for parallel checkpointing into step-92 or some such, where we can then write in the introduction about how checkpointing in parallel is different from sequential, and one addresses this.

I kind of like this approach because (i) it allows us to finish this program without having to write much more in the introduction, in particular about why we chose the `SolutionTransfer` approach for serialization, (ii) I like programs to be relatively incremental, and I think having one sequential and one parallel program is a nice progression in the same way as we have both step-6 and step-40, for example.

**Comment by kronbichler** [comment]
I agree that it would be great to wrap up what we have for the present release, so I am fine with whatever the participating people think is reasonable.

> [...]  I think having one sequential and one parallel program is a nice progression in the same way as we have both step-6 and step-40

Even though I did not read up into all details, Let me express a different opinion in this regard: I would prefer to integrate the parallel way of doing things after the release, if we cannot make it now. While I certainly see value in incremental progression, I also see that users get confused when there are several ways of doing things and too many places to look at. While the required adaptations can be seen as relatively straight-forward when one knows a substantial part of deal.II, less experienced users I have interacted with tend to pick one option without a complete understanding of what pieces they are putting together. From that perspective, a developer telling them "for parallel computations you should not use X, but switch to Y" can be met with the question "why do you show it the other way then as parallelism is the logical next step for my project?" Translating it into a more positive direction, I would think that given the effort we put into the documentation and tutorials, we might consider spending some of this effort on the interfaces that require as little effort as possible when making the sequential -> parallel transition. And sometimes showing less can be an advantage.

I say this even though I see myself guilty of providing several "options" in many places of the library. Not that options are bad in general, and we indeed also have highly qualified users who profit from them. Still, I cannot overstate how much I admire efforts like unifying `SolutionTransfer`, because they help almost uniformly.

**Comment by bangerth** [comment]
That's a reasonable point. But it seems like we're in favor of going with the route of least resistance for the moment, and getting this merged once #17266 is in place?

**Comment by blaisb** [comment]
I think this is a correct approach, however, I do feel the same way as @kronbichler regarding this.

An alternative could be to make a serialization example using step-68 for example which already uses parallel triangulations and solutions. This could be the "second" serialization example and it would have everything ( a triangulation, a DofHandler, particles and a solution vector). This is something I can take care of if you want. Step-68 is relatively trivial and writing a serialization of it for parallel::distributed::triangulation is straightforward.

**Comment by bangerth** [comment]
OK, then let's move forward with this and perhaps discuss the parallelization issue in a separate PR. I'll open that next.

**Comment by bangerth** [comment]
I've incorporated @peterrum 's suggestions now, and think this is ready for final review.

File: doc/doxygen/tutorial/tutorial.h.in
Comment (peterrum) [inline doc/doxygen/tutorial/tutorial.h.in:671]: We need other keywords!

File: examples/step-83/doc/intro.dox
Comment (peterrum) [inline examples/step-83/doc/intro.dox:225]: Which "this class"?

File: examples/step-83/doc/intro.dox
Comment (peterrum) [inline examples/step-83/doc/intro.dox:224]: The transition here is very abrupt. Until now, you talked generally about serialization/deserialization and suddenly you talk about the concrete example without introducing the concrete example.

File: examples/step-83/step-83.cc
Comment (peterrum) [inline examples/step-83/step-83.cc:755]: This is not correctly shown in doxygen:
```
std::ofstream checkpoint_file("tmp.checkpoint_@ref step_8 "step-8"3");
```

**Review by peterrum — APPROVED** [review]
The tutorial reads very well. I am fine with merging before release.

However, I feel very strongly about the usage of `SolutionTransfer` in the tutorial (once that PR is merged). We have/will have a nice interface that works both in serial and parallel, so why not use it!? All it would mean 1) that the vector is serialized/deserialized via `SolutionTransfer` and 2) this only happens after the data structures (in this case the vector) have been initialized. This is a small change to the tutorial but would mean that the checkpointing part of the tutorial would be straight forwardly parallelizable (with the exception that the files would need different names based on the ranks). Then there would be no need for a second tutorial, which would be great because, in this case, ALL checkpoint-related topics are encapsulated in a single tutorial and users don't need to switch between tutorials.

I have similar experiences with users as Martin has. 

> While the required adaptations can be seen as relatively straight-forward when one knows a substantial part of deal.II, less experienced users I have interacted with tend to pick one option without a complete understanding of what pieces they are putting together. From that perspective, a developer telling them "for parallel computations you should not use X, but switch to Y" can be met with the question "why do you show it the other way then as parallelism is the logical next step for my project?"

Many users start with serial tutorials. Once they need to parallelize it, they often major problems because there are just too many places one has to make "small" changes. I have seen multiple times instances, where it was easier to start with a parallel tutorial and to change that one so that it solves a different PDE. This is a situation that indicates that we need to do some efforts as library developers.

**Comment by blaisb** [comment]
I did not see it this way initially, but I believe @peterrum is right. One unified example that explains it in a more realistic fashion will prevent confusion in the long term. I think we should proceed this way. @peterrum did amazing work on the solution transfer unification. Why not leverage it ASAP?

**Comment by tamiko** [comment]
@bangerth @peterrum What is the state of this pull request?

**Comment by peterrum** [comment]
Good to merge once the test pass. Currently, `indent` is not passing.

**Review by tamiko — APPROVED** [review]

**Comment by bangerth** [comment]
I *think* I fixed and rebased this morning, but all testers seem to be pretty stuck today...

**Comment by bangerth** [comment]
Yay, thank you to everyone who helped with this program!

**Pull request merged**
